src/Datadog.Trace.ClrProfiler.Managed/Emit/MethodBuilder.cs

```diff
@@ -313,53 +313,79 @@ private TDelegate EmitDelegate()
                                          .ToArray();
             }
 
-            var dynamicMethod = Emit<TDelegate>.NewDynamicMethod(methodInfo.Name);
+            DynamicMethod dynamicMethod = new DynamicMethod(methodInfo.Name, returnType, delegateParameterTypes, ObjectExtensions.Module, skipVisibility: true);
+            ILGenerator il = dynamicMethod.GetILGenerator();
 
             // load each argument and cast or unbox as necessary
             for (ushort argumentIndex = 0; argumentIndex < delegateParameterTypes.Length; argumentIndex++)
             {
                 Type delegateParameterType = delegateParameterTypes[argumentIndex];
                 Type underlyingParameterType = effectiveParameterTypes[argumentIndex];
 
-                dynamicMethod.LoadArgument(argumentIndex);
+                switch (argumentIndex)
+                {
+                    case 0:
+                        il.Emit(OpCodes.Ldarg_0);
+                        break;
+                    case 1:
+                        il.Emit(OpCodes.Ldarg_1);
+                        break;
+                    case 2:
+                        il.Emit(OpCodes.Ldarg_2);
+                        break;
+                    case 3:
+                        il.Emit(OpCodes.Ldarg_3);
+                        break;
+                    default:
+                        il.Emit(OpCodes.Ldarg_S, argumentIndex);
+                        break;
+                }
 
                 if (underlyingParameterType.IsValueType && delegateParameterType == typeof(object))
                 {
-                    dynamicMethod.UnboxAny(underlyingParameterType);
+                    il.Emit(OpCodes.Unbox_Any, underlyingParameterType);
                 }
                 else if (underlyingParameterType != delegateParameterType)
                 {
-                    dynamicMethod.CastClass(underlyingParameterType);
+                    il.Emit(OpCodes.Castclass, underlyingParameterType);
                 }
             }
 
             if (_opCode == OpCodeValue.Call || methodInfo.IsStatic)
             {
                 // non-virtual call (e.g. static method, or method override calling overriden implementation)
-                dynamicMethod.Call(methodInfo);
+                il.Emit(OpCodes.Call, methodInfo);
             }
             else if (_opCode == OpCodeValue.Callvirt)
             {
                 // Note: C# compiler uses CALLVIRT for non-virtual
                 // instance methods to get the cheap null check
-                dynamicMethod.CallVirtual(methodInfo);
+                il.Emit(OpCodes.Callvirt, methodInfo);
             }
             else
             {
                 throw new NotSupportedException($"OpCode {_originalOpCodeValue} not supported when calling a method.");
             }
 
-            if (methodInfo.ReturnType.IsValueType && returnType == typeof(object))
+            if (methodInfo.ReturnType.IsValueType && !returnType.IsValueType)
```

Comment:
@tonyredondo I added more thorough error handling on the return types, please review

Reply:
looks great :shipit: